rosenpass/src/app_server.rs

```diff
@@ -584,6 +604,55 @@ impl AppServer {
         }
     }
 
+    fn handle_msg_under_load(
+        &mut self,
+        endpoint: &Endpoint,
+        rx: &[u8],
+        tx: &mut [u8],
+    ) -> Result<crate::protocol::HandleMsgResult> {
+        let socket_addr = endpoint
+            .addresses()
+            .first()
+            .ok_or(anyhow::anyhow!("No socket address for endpoint"))?;
+
+        //TODO: Modify AppPeer to store endpoints to enable O(1) search (hash) for addresses
+        //If no endpoint matches address, return error
+        //Currently, there is no sensible mechanism to map "anonymous" endpoints to PeerPtr
+        let index = self
+            .peers
+            .iter()
+            .enumerate()
+            .find(|(_num, p)| {
+                if let Some(ep) = p.endpoint() {
+                    ep.addresses().contains(socket_addr)
+                } else {
+                    false
+                }
+            })
+            .ok_or(anyhow::anyhow!("Received message from unknown endpoint"))?
+            .0;
+
+        let mut len = 0;
+        let mut ip_addr_port = [0u8; 18];
+
+        match socket_addr.ip() {
+            std::net::IpAddr::V4(ipv4) => {
+                ip_addr_port[0..4].copy_from_slice(&ipv4.octets());
+                len += 4;
+            }
+            std::net::IpAddr::V6(ipv6) => {
+                ip_addr_port[0..16].copy_from_slice(&ipv6.octets());
+                len += 16;
+            }
+        };
+
+        ip_addr_port[len..len + 2].copy_from_slice(&socket_addr.port().to_be_bytes());
+        len += 2;
+
+        self.crypt
+            .handle_msg_under_load(&rx[..len], &mut *tx, PeerPtr(index), &ip_addr_port[..len])
```

Comment:
Can we just have one handle_msg function which handles messages under load? This would also greatly simplify the `ReceivedMessage` case in `event_loop`